amethyst_renderer/src/pass/sprite/interleaved.rs

```diff
@@ -0,0 +1,149 @@
+    fn compile(&mut self, effect: NewEffect) -> Result<Effect> {
+        use std::mem;
+        let mut builder = effect.simple(VERT_SRC, FRAG_SRC);
+        builder
+            .with_raw_constant_buffer(
+                "VertexArgs",
+                mem::size_of::<<VertexArgs as Uniform>::Std140>(),
+                1,
+            )
+            .with_raw_constant_buffer(
+                "SpriteArgs",
+                mem::size_of::<<SpriteArgs as Uniform>::Std140>(),
+                1,
+            );
+        setup_textures(&mut builder, &TEXTURES);
+        match self.transparency {
+            Some((mask, blend, depth)) => builder.with_blended_output("color", mask, blend, depth),
+            None => builder.with_output("color", Some(DepthMode::LessEqualWrite)),
+        };
+        builder.build()
+    }
+
+    fn apply<'a, 'b: 'a>(
+        &'a mut self,
+        encoder: &mut Encoder,
+        effect: &mut Effect,
+        _factory: Factory,
+        (
+            active,
+            camera,
+            sprite_sheet_storage,
+            tex_storage,
+            material_texture_set,
+            visibility,
+            sprite_render,
+            global,
+        ): <Self as PassData<'a>>::Data,
+    ) {
+        let camera = get_camera(active, &camera, &global);
+
+        match visibility {
+            None => for (sprite_render, global) in (&sprite_render, &global).join() {
+                draw_sprite(
+                    encoder,
+                    effect,
+                    sprite_render,
+                    &sprite_sheet_storage,
+                    &tex_storage,
+                    &material_texture_set,
+                    camera,
+                    Some(global),
+                );
+            },
+            Some(ref visibility) => {
+                for (sprite_render, global, _) in
+                    (&sprite_render, &global, &visibility.visible_unordered).join()
+                {
+                    draw_sprite(
+                        encoder,
+                        effect,
+                        sprite_render,
+                        &sprite_sheet_storage,
+                        &tex_storage,
+                        &material_texture_set,
+                        camera,
+                        Some(global),
+                    );
+                }
+
+                for entity in &visibility.visible_ordered {
+                    if let Some(sprite_render) = sprite_render.get(*entity) {
+                        draw_sprite(
+                            encoder,
+                            effect,
+                            sprite_render,
+                            &sprite_sheet_storage,
+                            &tex_storage,
+                            &material_texture_set,
+                            camera,
+                            global.get(*entity),
+                        );
+                    }
+                }
+            }
```

Comment:
I see this piece of code over and over again in passes. Any way to put that in utils.rs instead?

Reply:
the repetition is in the `pbm`, `shaded`, and `flat` passes which I haven't touched; this sprite pass is notably different to those in the sense that it:

* doesn't use `Material` or `Mesh`
* doesn't use non-uniform vertex buffers
* uses sprite related components / storages

if it were to be commonized, perhaps it should be in a separate PR so that something that says "implement sprites pass" doesn't inadvertently break other rendering passes unrelated to sprites

Reply:
There's no way to do that, because types are different, and have no commonality. This was the best that could be done for unification of the passes.